Serve whitehall organisation logos from asset manager

chrisroos opened the pull request:
__NOTE.__ This can't be merged until the following dependencies have been deployed:

* [Upload all existing Whitehall organisation logos to Asset Manager](https://github.com/alphagov/asset-manager/issues/276)
  * Merging this change first would result in 404s for organisation logos that don't yet exist on Asset Manager.
* [Configure Asset Manager to serve Whitehall assets from S3](https://github.com/alphagov/asset-manager/issues/283)
  * Merging this PR first would mean we'd invalidate the caches of existing Whitehall organisation logos. While this isn't a big deal in the case of organisations (because there are so few of them) it would potentially cause us problems when migrating a larger group of assets.

---

This routes requests for Whitehall organisations logos to Asset Manager instead of to Whitehall. See [this milestone](https://github.com/alphagov/asset-manager/milestone/9) for more info.

The change to `common.yaml` results in a new `location` block being added to the `assets-origin` Nginx config to proxy the request to `static`.

The change to `static_extra_nginx_config.conf.erb` adds a `location` block to proxy the request to `asset-manager`. The additional configuration is a duplication of the configuration required to proxy requests for `/media` to `static` in the same file. I plan to DRY this duplication up in a subsequent commit.

Requests for organisation logos were previously being proxied from `assets-origin` to `whitehall-frontend` by the `/government/uploads` location configured in `hierdata/common.yaml`.

I've tested the effect of this change on my development VM and am confident that it's working as expected. I saw the following differences in response headers but I don't think we need to worry about any of them:

* Cache-Control
  * `max-age=14400, public` from Whitehall
  * `private` from Asset Manager
  * This is expected because I'm running Asset Manager in development mode. It'll add the correct `Cache-Control` headers in production.

* Last-Modified
  * `Wed, 25 Oct 2017 15:01:08 GMT` from Whitehall
  * `Wed, 25 Oct 2017 15:00:34 GMT` from Asset Manager
  * This is expected because this was a new image uploaded to Whitehall which was then sent to Asset Manager so it'll have been written to the two different places on disk at different times.

* ETag
  * `"59f0a734-13767"` from Whitehall
  * `"59f0a712-13767"` from Asset Manager
  * This is expected because this was a new image uploaded to Whitehall which was then sent to Asset Manager. Part of this etag is based on the Last Modified time which, as explained above, is also different.

* Strict-Transport-Security
  * Not present from Whitehall
  * `max-age=31536000` from Asset Manager
  * This is added by the `location` block in the static nginx configuration.
  * It seems fine to add this header given that we want to encourage browsers to always access this site over HTTPS.

floehopper commented on modules/govuk/templates/static_extra_nginx_config.conf.erb:
```diff
@@ -29,3 +29,17 @@ location ~ ^/media/ {
+  # forces nginx not to clear Cache-Control headers further up the
+  # stack.
+  include /etc/nginx/add-sts.conf;
+}
```
> * Strict-Transport-Security
>   * Not present from Whitehall

Although this header is *not* present in responses served by Whitehall on the development VM, I believe [it *is* (or at least was) present in responses served by Whitehall in production][1].

[1]: https://github.com/alphagov/asset-manager/issues/236#issuecomment-333109045

chrisroos commented on hieradata/common.yaml:
```diff
@@ -1485,6 +1485,7 @@ router::assets_origin::asset_routes:
   '/government/assets/': "whitehall-frontend"
   '/government/placeholder': "whitehall-frontend"
   '/government/uploads/': "whitehall-frontend"
+  '/government/uploads/system/uploads/organisation/logo': "static"
```
> Some of the path patterns have a trailing forward-slash, but others do not. Do you know why? And are you confident that having no trailing forward-slash is correct?

I've added a trailing slash to this new path but not looked at the other paths to know whether they should or shouldn't have trailing slashes. The lack of trailing slash meant that, for example, `.../logo-foo` would've been matched by this new location. Adding the trailing slash prevents this and ensures that the trailing slash has to be present in the request URI. In practice, I'm not sure it'll make any real difference but I think adding the slash makes the intention clearer. Good spot!

chrisroos replied:
> Given you've tested this on your development VM I assume the router matches this new route in preference to '/government/uploads/'? Do you know how come this work? Is it because the router matches longer routes in preference to shorter ones? I think it might be worth explaining this in the commit note.

I'm fairly certain that requests for these assets don't go through the router - they just go through nginx. As such, it's the nginx behaviour that prefers a longer match to a shorter match when calculating the location block to use. I've updated the commit message to explain this:

> Note that Nginx matches longer prefix matches in preference to shorter
> matches[1], which is why requests for organisations logos will be
> handled by our new location block instead of the `/government/uploads/`
> block that was already in place.

chrisroos commented on modules/govuk/templates/static_extra_nginx_config.conf.erb:
```diff
@@ -16,30 +16,22 @@ location /robots.txt {
-  proxy_set_header X-Forwarded-Host $host;
-  proxy_pass <%= @enable_ssl ? 'https' : 'http' %>://<%= @asset_manager_host %>;
+<% [
+  '/media',
```
Good spot! It wasn't intentional. I've amended the commit to add it back.

chrisroos commented on modules/govuk/templates/static_extra_nginx_config.conf.erb:
```diff
@@ -16,30 +16,22 @@ location /robots.txt {
+<% [
+  '/media',
+  '/government/uploads/system/uploads/organisation/logo/'
+].each do |location_to_proxy| %>
```
I've amended the commit to rename it to `path_to_be_proxied_to_asset_manager`.

chrisroos commented:
Thanks for reviewing, @floehopper. I've addressed your comments and so I think this is ready to merge once we've uploaded the existing organisation logos.

chrisroos commented:
The two prerequisites for this PR have now been completed so I'm going to get this merged!
